engine/access/rpc/backend/backend_stream_transactions.go

```diff
@@ -0,0 +1,181 @@
+// subscription responses based on new blocks.
+func (b *backendSubscribeTransactions) getTransactionStatusResponse(txInfo *TransactionSubscriptionMetadata) func(context.Context, uint64) (interface{}, error) {
+	return func(ctx context.Context, height uint64) (interface{}, error) {
+		highestHeight, err := b.blockTracker.GetHighestHeight(flow.BlockStatusFinalized)
+		if err != nil {
+			return nil, fmt.Errorf("could not get highest height for block %d: %w", height, err)
+		}
+
+		// Fail early if no block finalized notification has been received for the given height.
+		// Note: It's possible that the block is locally finalized before the notification is
+		// received. This ensures a consistent view is available to all streams.
+		if height > highestHeight {
+			return nil, fmt.Errorf("block %d is not available yet: %w", height, storage.ErrNotFound)
+		}
+
+		if txInfo.blockWithTx == nil {
+			// Check if block contains transaction.
+			txInfo.blockWithTx, err = b.searchForTransactionBlock(height, txInfo)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Find the transaction status.
+		var txStatus flow.TransactionStatus
+		if txInfo.blockWithTx == nil {
+			txStatus, err = b.txLocalDataProvider.DeriveUnknownTransactionStatus(txInfo.txReferenceBlockID)
+		} else {
+			blockID := txInfo.blockWithTx.ID()
+
+			if !txInfo.txExecuted {
+				// Check if transaction was executed.
+				txInfo.txExecuted, err = b.searchForExecutionResult(blockID)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			txStatus, err = b.txLocalDataProvider.DeriveTransactionStatus(blockID, txInfo.blockWithTx.Height, txInfo.txExecuted)
+		}
+		if err != nil {
+			if !errors.Is(err, state.ErrUnknownSnapshotReference) {
+				irrecoverable.Throw(ctx, err)
+			}
+			return nil, rpc.ConvertStorageError(err)
+		}
+
+		messageIndex := txInfo.messageIndex.Value()
+		if ok := txInfo.messageIndex.Set(messageIndex + 1); !ok {
+			return nil, status.Errorf(codes.Internal, "the message index has already been incremented to %d", txInfo.messageIndex.Value())
+		}
+
+		return &convert.TransactionSubscribeInfo{
+			ID:           txInfo.txID,
+			Status:       txStatus,
+			MessageIndex: messageIndex,
+		}, nil
+	}
+}
+
+// searchForTransactionBlock searches for the block containing the specified transaction.
+// It retrieves the block at the given height and checks if the transaction is included in that block.
+// Expected errors:
+// - codes.Internal when unable to retrieve the block or collection ID
+func (b *backendSubscribeTransactions) searchForTransactionBlock(
+	height uint64,
+	txInfo *TransactionSubscriptionMetadata,
+) (*flow.Header, error) {
+	block, err := b.txLocalDataProvider.blocks.ByHeight(height)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "could not get block %d: %v", height, err)
+	}
+
+	collectionID, err := b.txLocalDataProvider.LookupCollectionIDInBlock(block, txInfo.txID)
```

Comment:
this can return errors in 2 cases:
* `storage.ErrNotFound` when the collection is found in storage -> indicates block has ben finalized locally, but the collection has not been fetched from the network yet
* `codes.NotFound` when the tx is not included in any collection within the block

I think we should handle the errors here a little differently. My suggestion is to update the case where the tx is not in the block to use a new sentinel error, `ErrTransactionNotInBlock` or something like that. This can be used to identify the specific case that the block is fully indexed, and the tx just doesn't exist. Then in the case that the collection isn't found, just return the error directly.

There would then be 2 expected sentinel errors for  `LookupCollectionIDInBlock`,
- `ErrTransactionNotInBlock` -> tx is not included in block
- `storage.ErrNotFound` -> one of the collections contained in the block is not available.

From there, we can update this method to check for both of these sentinels, and handle them:
- `ErrTransactionNotInBlock` -> handled the way `codes.NotFound` is here
- `storage.ErrNotFound` -> handled by returning an error instructing the streamer to retry this height later
- Any other error returns `codes.Internal`